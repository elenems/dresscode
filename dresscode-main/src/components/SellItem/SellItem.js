import React, { Component } from "react";
import axios from "axios";
import TextInput from "../TextInput/TextInput";
import Select from "../Select/Select";
import PropTypes from "prop-types";

const conditionOptions = [
  { value: "New", label: "New" },
  { value: "Perfect", label: "Perfect" },
  { value: "Good", label: "Good" },
  { value: "Average", label: "Average" }
];

const categoryOptions = [
  { value: "Tops", label: "Tops" },
  { value: "Bottoms", label: "Bottoms" },
  { value: "Outwear", label: "Outwear" },
  { value: "Boots", label: "Boots" },
  { value: "Sneakers", label: "Sneakers" },
  { value: "Hats", label: "Hats" },
  { value: "Accessories", label: "Accessories" }
];
const MAX_FILE_SIZE = 3000 * 1048; // 3MB
// description, price, category, title, condition, | brand, views, addedDate, likes, addedBy, comments
class SellItem extends Component {
  state = {
    Price: "0",
    Condition: { value: "New", label: "New" },
    Description: "",
    Category: { value: "Tops", label: "Tops" },
    Title: "",
    Brand: "",
    Size: "",
    fileMessage: "",
    statusInfo: "",
    statusColor: "",
    selectedFile: null,
    errors: {
      titleError: "",
      brandError: "",
      conditionError: "",
      categoryError: ""
    }
  };

  handleChange = e => {
    this.setState({
      [e.target.id]: e.target.value
    });
  };

  onChangeHandler = event => {
    this.setState({
      selectedFile: event.target.files[0],
      message: ""
    });
  };

  onClickHandler = e => {
    e.preventDefault();
    if (this.state.selectedFile === null) {
      this.setState({
        fileMessage: "Please choose a photo"
      });
    } else {
      this.setState({
        fileMessage: ""
      });
      if (this.state.selectedFile.size > MAX_FILE_SIZE) {
        this.setState({ fileMessage: "File to big" });
      } else if (
        this.state.selectedFile.type === "image/jpeg" ||
        this.state.selectedFile === "image/png"
      ) {
        const data = new FormData();
        data.append("file", this.state.selectedFile);
        const {
          Title,
          Brand,
          Description,
          Price,
          Condition,
          Category,
          Size
        } = this.state;
        const link = `/addClothes?title=${Title}&addedBy=${
          this.props.user.user.userId
        }&size=${Size}&brand=${Brand}&description=${Description}&price=${Price}&condition=${
          Condition.value
        }&category=${Category.value}`;
        axios
          .post(link, data)
          .then(data => {
            window.scrollTo({
              top: 0,
              left: 0,
              behavior: "smooth"
            });
            this.setState({
              fileMessage: "",
              statusColor: "#19b719",
              statusInfo: "Successfuly added. You are free to add a new one.",
              Price: "0",
              Condition: { value: "new", label: "New" },
              Description: "",
              Category: { value: "tops", label: "Tops" },
              Title: "",
              Brand: "",
              Size: "",
              selectedFile: null,
              errors: {
                titleError: "",
                brandError: "",
                conditionError: "",
                categoryError: ""
              }
            });
            document.getElementById("sell-form").reset();
            setTimeout(() => {
              this.setState({ statusInfo: "" });
            }, 3000);
          })
          .catch(e => {
            window.scrollTo({
              top: 0,
              left: 0,
              behavior: "smooth"
            });
            this.setState({
              errors: e.response.data.errors,
              statusColor: "#de3650",
              statusInfo: "Error adding new clothes"
            });
            setTimeout(() => {
              this.setState({ statusInfo: "" });
            }, 3000);
          });
      } else {
        this.setState({ fileMessage: "Upload image with jpg/png type" });
      }
    }
  };

  handleCategoryChange = selectedOption => {
    this.setState({ Category: selectedOption });
  };

  handleConditionChange = selectedOption => {
    this.setState({ Condition: selectedOption });
  };

  render() {
    return (
      <div className="sell-item-block row">
        <div className="module">
          {this.state.statusInfo ? (
            <p className="message" style={{ color: this.state.statusColor }}>
              {this.state.statusInfo}
            </p>
          ) : null}
          <h1 className="text-center">Add a new listing</h1>
          <form id="sell-form" className="sell-form">
            <TextInput
              handleChange={this.handleChange}
              value={this.state.Title}
              place="Clothes title"
              name="Title"
              type="text"
              error={this.state.errors.titleError}
            />

            <TextInput
              handleChange={this.handleChange}
              value={this.state.Brand}
              place="Enter clothes brand"
              name="Brand"
              type="text"
              error={this.state.errors.brandError}
            />

            <TextInput
              handleChange={this.handleChange}
              value={this.state.Price}
              place="Enter price $"
              name="Price"
              type="text"
              error=""
            />

            <TextInput
              handleChange={this.handleChange}
              value={this.state.Size}
              place="Enter size"
              name="Size"
              type="text"
              error=""
            />

            <div className="input-container">
              <Select
                label="Clothes condition"
                value={this.state.Condition}
                handleChange={this.handleConditionChange}
                options={conditionOptions}
              />
            </div>

            <div className="input-container">
              <Select
                label="Choose a category"
                value={this.state.Category}
                handleChange={this.handleCategoryChange}
                options={categoryOptions}
              />
            </div>

            <div className="input-container">
              <textarea
                onChange={this.handleChange}
                value={this.state.Description}
                id="Description"
                placeholder="Description"
              />
            </div>

            <div className="input-container">
              <span className="label">Choose a photo</span>
              <input type="file" name="file" onChange={this.onChangeHandler} />
              <p style={{ color: "red" }}>{this.state.fileMessage}</p>
            </div>

            <button
              onClick={this.onClickHandler}
              className="primary-button primary-button--big"
            >
              Add
            </button>
          </form>
        </div>
      </div>
    );
  }
}

SellItem.propTypes = {
  user: PropTypes.object
};

export default SellItem;
