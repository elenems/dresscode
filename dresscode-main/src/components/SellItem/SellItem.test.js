import React from "react";
import { shallow } from "enzyme";
import SellItem from "./SellItem";
import axios from 'axios';
describe("SellItem", () => {
  beforeAll(()=>{
    axios.defaults.baseURL =
    "https://europe-west2-dresscode-691e5.cloudfunctions.net/api";
  })
  const props = {
    user: {
      user: {
        userId: 'NMkib9WrLZBjSOFl6jSG'
      }
    }
  };

  it("Matches snapshot", () => {
    const wrapper = shallow(<SellItem {...props} />).debug();
    expect(wrapper).toMatchSnapshot();
  });

  it("Shows status info", () => {
    const wrapper = shallow(<SellItem {...props} />);
    wrapper.setState({
      statusInfo: "success",
      statusColor: "#fff"
    });
    expect(wrapper.find(".message").text()).toBe("success");
    expect(wrapper.find(".message").prop("style")).toEqual({
      color: "#fff"
    });
  });

  it("Changes condition", () => {
    const wrapper = shallow(<SellItem {...props} />);
    wrapper.instance().handleConditionChange("new");
    expect(wrapper.state("Condition")).toBe("new");
  });

  it("Changes category", () => {
    const wrapper = shallow(<SellItem {...props} />);
    wrapper.instance().handleCategoryChange("boots");
    expect(wrapper.state("Category")).toBe("boots");
  });

  it("Handles changes", () => {
    const event = {
      target: {
        id: "Size",
        value: "L"
      }
    };
    const wrapper = shallow(<SellItem {...props} />);
    wrapper.instance().handleChange(event);
    expect(wrapper.state("Size")).toBe("L");
  });

  it("Handles file change", () => {
    const event = {
      target: {
        files: [{ name: "image" }]
      }
    };
    const wrapper = shallow(<SellItem {...props} />);
    wrapper.instance().onChangeHandler(event);
    expect(wrapper.state("selectedFile")).toEqual({ name: "image" });
  });

  it("Handles onClick", () => {
    const mock = jest.fn();
    const event = {
      preventDefault: () => {
        mock();
      }
    };
    const wrapper = shallow(<SellItem {...props} />);
    wrapper.instance().onClickHandler(event);
    expect(mock).toHaveBeenCalled();
  });

  it("Adds file message if image is null", () => {
    const event = { preventDefault: () => {} };
    const wrapper = shallow(<SellItem {...props} />);
    wrapper.instance().onClickHandler(event);
    expect(wrapper.state("fileMessage")).toBe("Please choose a photo");
  });

  it("Add file message if image size is to big", () => {
    const event = { preventDefault: () => {} };
    const wrapper = shallow(<SellItem {...props} />);
    const file = {
      size: 3000123123123
    };
    wrapper.setState({
      selectedFile: file
    });
    wrapper.instance().onClickHandler(event);
    expect(wrapper.state("fileMessage")).toBe("File to big");
  });

  it("Add file message if image type is wrong", () => {
    const event = { preventDefault: () => {} };
    const wrapper = shallow(<SellItem {...props} />);
    const file = {
      size: 30001,
      type:'wrong type'
    };
    wrapper.setState({
      selectedFile: file
    });
    wrapper.instance().onClickHandler(event);
    expect(wrapper.state("fileMessage")).toBe("Upload image with jpg/png type");
  });

//   it("", async () => {
//     const event = { preventDefault: () => {} };
//     const wrapper = shallow(<SellItem {...props} />);
//     window.scrollTo=()=>{}
//     console.log(new File(['(⌐□_□)'], 'chucknorris.png', { type: 'image/png', size:23 }))
//     wrapper.findWhere(n=>n.prop('type') === 'file').simulate('change', {
//         target: {
//            files: [
//             new File(['(⌐□_□)'], 'chucknorris.png', { type: 'image/png', size:23 })
//            ]   
//         }
//       });
//     const data = await wrapper.instance().onClickHandler(event);
//     // expect(wrapper.state("fileMessage")).toBe("Upload image with jpg/png type");
//   });
});
