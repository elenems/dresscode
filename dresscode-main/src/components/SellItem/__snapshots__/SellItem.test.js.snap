// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SellItem Matches snapshot 1`] = `
"<div className=\\"sell-item-block row\\">
  <div className=\\"module\\">
    <h1 className=\\"text-center\\">
      Add a new listing
    </h1>
    <form id=\\"sell-form\\" className=\\"sell-form\\">
      <Memo(TextInput) handleChange={[Function]} value=\\"\\" place=\\"Clothes title\\" name=\\"Title\\" type=\\"text\\" error=\\"\\" />
      <Memo(TextInput) handleChange={[Function]} value=\\"\\" place=\\"Enter clothes brand\\" name=\\"Brand\\" type=\\"text\\" error=\\"\\" />
      <Memo(TextInput) handleChange={[Function]} value=\\"0\\" place=\\"Enter price $\\" name=\\"Price\\" type=\\"text\\" error=\\"\\" />
      <Memo(TextInput) handleChange={[Function]} value=\\"\\" place=\\"Enter size\\" name=\\"Size\\" type=\\"text\\" error=\\"\\" />
      <div className=\\"input-container\\">
        <Memo(Select) label=\\"Clothes condition\\" value={{...}} handleChange={[Function]} options={{...}} />
      </div>
      <div className=\\"input-container\\">
        <Memo(Select) label=\\"Choose a category\\" value={{...}} handleChange={[Function]} options={{...}} />
      </div>
      <div className=\\"input-container\\">
        <textarea onChange={[Function]} value=\\"\\" id=\\"Description\\" placeholder=\\"Description\\" />
      </div>
      <div className=\\"input-container\\">
        <span className=\\"label\\">
          Choose a photo
        </span>
        <input type=\\"file\\" name=\\"file\\" onChange={[Function]} />
        <p style={{...}} />
      </div>
      <button onClick={[Function]} className=\\"primary-button primary-button--big\\">
        Add
      </button>
    </form>
  </div>
</div>"
`;
